src/main/kotlin/org/rust/ide/highlight/RsHighlightExitPointsHandlerFactory.kt

```diff
@@ -51,11 +55,18 @@ private class RsHighlightExitPointsHandler(editor: Editor, file: PsiFile, var ta
             }
         }
 
-        val fnOrLambda = target.ancestors.firstOrNull { it is RsFunction || it is RsLambdaExpr }
-        when (fnOrLambda) {
-            is RsFunction -> ExitPoint.process(fnOrLambda, sink)
-            is RsLambdaExpr -> ExitPoint.process(fnOrLambda, sink)
+        loop@ for (it in target.ancestors) {
+            when {
+                it is RsBlockExpr
+                    && it.isTry
+                    && target.elementType == Q -> Unit
+                it is RsBlockExpr
+                    && it.isAsync -> ExitPoint.process(it.block, sink)
```

Comment:
Currently, there are no tests for this case (`return` highlighting inside `async { ... }`). Please add it. 

Also, the code can be rewritten in a more readable and coverage-friendly way (and then code coverage will show that `async` case is uncovered), like this:
```Rust
for (ancestor in target.ancestors) {
    if (ancestor is RsBlockExpr && ancestor.isTry && target.elementType == Q) {
        break
    } else if (ancestor is RsBlockExpr && ancestor.isAsync) {
        ExitPoint.process(ancestor.block, sink)
        break
    } else if (ancestor is RsFunction) {
        ExitPoint.process(ancestor.block, sink)
        break
    } else if (ancestor is RsLambdaExpr) {
        ExitPoint.process(ancestor.expr, sink)
        break
    }
}
```